### PR TITLE
Dropdown impl

### PR DESCRIPTION
**Pull request by amiryonatan**

### CODE REVIEW

File: packages/wix-ui-core/src/components/Dropdown/Dropdown.tsx [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+  render() {
+    const {openTrigger, placement, options, children} = this.props;
+    const {isOpen} = this.state;
+
+    return (
+      <Popover
+        placement={placement}
+        shown={isOpen}
+        onMouseEnter={openTrigger === HOVER ? () => this.setState({isOpen: true}) : null}
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/Dropdown/Dropdown.tsx]: add close and open funcs for this.setState({isOpen: true})

File: packages/wix-ui-core/src/components/Dropdown/Dropdown.tsx [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        onMouseLeave={openTrigger === HOVER ? () => this.setState({isOpen: false}) : null}>
+        <Popover.Element>
+          <div
+            data-hook="dropdown-element"
+            onClick={openTrigger === CLICK ? () => this.setState({isOpen: true}) : null}
+            style={{display: 'inline-block'}}>
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/Dropdown/Dropdown.tsx]: ???

File: packages/wix-ui-core/src/utils/index.ts [hunk 3]
```diff
@@ -22,3 +22,9 @@ export const createComponentThatRendersItsChildren = displayName => {
   Element.displayName = displayName;
   return Element;
 };
+
+export const getRandomInt = (min, max) =>  {
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/utils/index.ts]: remove. lodash.uniqueId()

File: packages/wix-ui-core/stories/index.tsx [hunk 4]
```diff
@@ -17,6 +27,28 @@ storiesOf('Components', module)
   .add('Button', () => (
     <Button dataHook="story-button">Hello</Button>
   ))
+  .add('Dropdown', () => (
```
Comment (lbelinsk) [inline packages/wix-ui-core/stories/index.tsx]: remove

File: packages/wix-ui-core/stories/index.tsx [hunk 5]
```diff
@@ -17,6 +27,28 @@ storiesOf('Components', module)
...
+      </Dropdown>
+      <div style={{display: 'inline-block', width: '20px'}} />
+      <Dropdown dataHook="story-dropdown" openTrigger="hover" mode="multiSelect" selectedIds={[1, 2, 6]} options={dropdownOptions}>
+        {({selectedOptions}) => <span>{selectedOptions.map(x => x.displayName).join() || 'Select option'}</span>}
+      </Dropdown>
+    </div>
+  ))
+  .add('IconWithOptions', () => (
+    <div style={{padding: '50px'}}>
+      <input tabIndex={1} />
```
Comment (lbelinsk) [inline packages/wix-ui-core/stories/index.tsx]: tabIbdex remove?

File: packages/wix-ui-core/stories/index.tsx [hunk 6]
```diff
@@ -17,6 +27,28 @@ storiesOf('Components', module)
...
+      <IconWithOptions
+        tabIndex={2}
+        iconUrl="https://cdn3.iconfinder.com/data/icons/caps-hats/512/Ladies_cap-128.png"
+        mode="multiSelect"
```
Comment (lbelinsk) [inline packages/wix-ui-core/stories/index.tsx]: add to your "not imaginary" doc

File: packages/wix-ui-core/src/components/IconWithOptions/IconWithOptions.tsx [hunk 7]
```diff
@@ -0,0 +1,56 @@
+import * as React from 'react';
+import Dropdown from '../Dropdown';
+import {SharedDropdownProps} from '../Dropdown/Dropdown';
+import {HOVER, CLICK, SINGLE_SELECT, MULTI_SELECT} from '../Dropdown/Dropdown';
+import {createHOC} from '../../createHOC';
+import {oneOf, string, object, func, arrayOf, number} from 'prop-types';
+
+export interface IconWithOptionsProps extends SharedDropdownProps {
+  iconUrl: string;
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/IconWithOptions/IconWithOptions.tsx]: to be react component , so write it down to refactor later

File: packages/wix-ui-core/src/components/IconWithOptions/IconWithOptions.tsx [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+  tabIndex?: number;
+}
+
+const IconWithOptions: React.SFC<IconWithOptionsProps> = props => {
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/IconWithOptions/IconWithOptions.tsx]: destructuring

File: packages/wix-ui-core/src/components/IconWithOptions/IconWithOptions.tsx [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+  return (
+    <Dropdown
+      options={options}
+      placement={placement}
+      openTrigger={openTrigger}
+      mode={mode}
+      onSelect={onSelect}>
+      {
+        () =>
+          <img
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/IconWithOptions/IconWithOptions.tsx]: tmp!

File: packages/wix-ui-core/src/components/IconWithOptions/IconWithOptions.tsx [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+            src={iconUrl}
+            tabIndex={tabIndex}
+            onKeyDown={(evt: React.KeyboardEvent<HTMLImageElement>) => console.log(evt.keyCode)}/>
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/IconWithOptions/IconWithOptions.tsx]: console.log

File: packages/wix-ui-core/src/components/Dropdown/DropdownContent.tsx [hunk 11]
```diff
@@ -0,0 +1,78 @@
+import * as React from 'react';
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/Dropdown/DropdownContent.tsx]: Reconsider folder structure. Maybe base components?

File: packages/wix-ui-core/src/components/Dropdown/DropdownContent.tsx [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+
+  constructor(props) {
+    super(props);
+
+    this._renderOption = this._renderOption.bind(this);
+  }
+
+  _onOptionClick(option, evt) {
+    const {onOptionClick} = this.props;
+    onOptionClick && onOptionClick(option, evt);
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/Dropdown/DropdownContent.tsx]: noop default

File: packages/wix-ui-core/src/components/Dropdown/Dropdown.spec.tsx [hunk 13]
```diff
@@ -0,0 +1,145 @@
+import * as React from 'react';
+import {createDriverFactory} from 'wix-ui-test-utils';
+import {dropdownDriverFactory} from './Dropdown.driver';
```
[the remark this answers is not in the document]
Comment (amiryonatan) [reply packages/wix-ui-core/src/components/Dropdown/Dropdown.spec.tsx]: 👍

File: packages/wix-ui-core/src/components/Dropdown/Dropdown.spec.tsx [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+
+  describe('openTrigger', () => {
+    it ('should show content on click', () => {
+      const driver = createDriver(createDropdown());
+
+      driver.click();
+      expect(driver.isContentElementExists()).toBeTruthy();
+    });
+
+    it ('should show content on on hover', () => {
```
[the remark this answers is not in the document]
Comment (amiryonatan) [reply packages/wix-ui-core/src/components/Dropdown/Dropdown.spec.tsx]: 👍

File: packages/wix-ui-core/src/components/Dropdown/Dropdown.spec.tsx [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+
+    const options = [1, 2, 3, 4, 5].map(x => ({
+      id: x,
+      value: `value${x}`,
+      displayName: `value ${x}`,
+      type: x === 3 ? 'separator' : 'option',
+      isDisabled: x === 4
+    }));
+
+  it ('should render dropdown default dropdown', () => {
```
[the remark this answers is not in the document]
Comment (amiryonatan) [reply packages/wix-ui-core/src/components/Dropdown/Dropdown.spec.tsx]: 👍

File: packages/wix-ui-core/src/components/Dropdown/Dropdown.tsx (hunk 1 above)
Comment (amiryonatan) [reply packages/wix-ui-core/src/components/Dropdown/Dropdown.tsx]: 👍

File: packages/wix-ui-core/src/components/Dropdown/Dropdown.tsx (hunk 2 above)
Comment (amiryonatan) [reply packages/wix-ui-core/src/components/Dropdown/Dropdown.tsx]: 👍

File: packages/wix-ui-core/src/components/Dropdown/Dropdown.spec.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (lbelinsk) [reply packages/wix-ui-core/src/components/Dropdown/Dropdown.spec.tsx]: OOPs, I see that I was mistaken, no default in other components

File: packages/wix-ui-core/src/components/Dropdown/Dropdown.spec.tsx [hunk 16]
```diff
@@ -0,0 +1,145 @@
+import * as React from 'react';
+import {createDriverFactory} from 'wix-ui-test-utils';
+import {dropdownDriverFactory} from './Dropdown.driver';
+import Dropdown from './index';
+import {HOVER, MULTI_SELECT} from './Dropdown';
```
[the remark this answers is not in the document]
Comment (amiryonatan) [reply packages/wix-ui-core/src/components/Dropdown/Dropdown.spec.tsx]: 👍

File: packages/wix-ui-core/src/components/Dropdown/Dropdown.tsx [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+    const isSingleSelect = this.isSingleSelect();
+    const {onSelect, onDeselect} = this.props;
+    const newState = {
+      isOpen: !isSingleSelect,
+      selectedOptions: []
+    };
+
+    if (isSingleSelect) {
+      newState.selectedOptions = [option];
+      onSelect && onSelect(option, evt);
```
[the remark this answers is not in the document]
Comment (amiryonatan) [reply packages/wix-ui-core/src/components/Dropdown/Dropdown.tsx]: 👍

File: packages/wix-ui-core/src/components/Dropdown/DropdownContent.tsx (hunk 12 above)
Comment (amiryonatan) [reply packages/wix-ui-core/src/components/Dropdown/DropdownContent.tsx]: 👍

File: packages/wix-ui-core/src/components/IconWithOptions/IconWithOptions.tsx (hunk 8 above)
Comment (amiryonatan) [reply packages/wix-ui-core/src/components/IconWithOptions/IconWithOptions.tsx]: 👍

File: packages/wix-ui-core/src/components/Dropdown/Dropdown.spec.tsx [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+      driver.click();
+      driver.clickOptionAt(0);
+      expect(driver.targetText()).toEqual('value 1');
+
+      driver.click();
+      driver.clickOptionAt(1);
+      expect(driver.targetText()).toEqual('value 2');
+
+      driver.click();
+      driver.clickOptionAt(2); // Separator, do nothing
```
[the remark this answers is not in the document]
Comment (amiryonatan) [reply packages/wix-ui-core/src/components/Dropdown/Dropdown.spec.tsx]: 👍

File: packages/wix-ui-core/src/components/Dropdown/Dropdown.spec.tsx [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+      const driver = createDriver(createDropdown({options, selectedId: null}));
+      expect(driver.targetText()).toEqual('Select option');
+    });
+
+    it('should display selected option with initial value', () => {
+      const driver = createDriver(createDropdown({options, selectedId: 5}));
+      expect(driver.targetText()).toEqual('value 5');
+    });
+
+    it('should display selected option with initial invalid value', () => {
```
[the remark this answers is not in the document]
Comment (amiryonatan) [reply packages/wix-ui-core/src/components/Dropdown/Dropdown.spec.tsx]: 👍

File: packages/wix-ui-core/src/components/Dropdown/Dropdown.spec.tsx [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+    it('should display selected options with initial value', () => {
+      const driver = createDriver(createDropdown({options, selectedIds: [1, 5]}));
+      expect(driver.targetText()).toEqual('value 1,value 5');
+    });
+
+    it('should display selected options with initial value one invalid index', () => {
+      const driver = createDriver(createDropdown({options, selectedIds: [1, 6]}));
+      expect(driver.targetText()).toEqual('value 1');
+    });
+
```
[the remark this answers is not in the document]
Comment (amiryonatan) [reply packages/wix-ui-core/src/components/Dropdown/Dropdown.spec.tsx]: 👍

File: packages/wix-ui-core/src/components/Dropdown/Dropdown.spec.tsx [hunk 21]
```diff
@@ -0,0 +1,145 @@
...
+  describe('onSelect', () => {
+    it('should be called when selection changed', () => {
+      const onSelect = jest.fn();
+      const driver = createDriver(createDropdown({options, onSelect}));
+
+      driver.click();
+      driver.clickOptionAt(0);
+      expect(onSelect).toHaveBeenCalled();
+    });
```
[the remark this answers is not in the document]
Comment (amiryonatan) [reply packages/wix-ui-core/src/components/Dropdown/Dropdown.spec.tsx]: 👍

File: packages/wix-ui-core/src/components/IconWithOptions/IconWithOptions.tsx (hunk 10 above)
Comment (amiryonatan) [reply packages/wix-ui-core/src/components/IconWithOptions/IconWithOptions.tsx]: 👍

File: packages/wix-ui-core/stories/index.tsx (hunk 4 above)
Comment (amiryonatan) [reply packages/wix-ui-core/stories/index.tsx]: 👍

File: packages/wix-ui-core/stories/index.tsx (hunk 5 above)
Comment (amiryonatan) [reply packages/wix-ui-core/stories/index.tsx]: 👍

File: packages/wix-ui-core/src/components/Dropdown/Dropdown.driver.ts [hunk 22]
```diff
@@ -0,0 +1,14 @@
...
+const getOptionAt = (element, index) => element.querySelectorAll('[data-hook="options-container"] > div')[index];
+
+export const dropdownDriverFactory = ({element, eventTrigger}) => ({
+  exists: () => !!element,
+  isTargetElementExists: () => !!getElement(element),
+  isContentElementExists: () => !!getContent(element),
+  mouseEnter: () => eventTrigger.mouseEnter(element),
+  mouseLeave: () => eventTrigger.mouseLeave(element),
+  click: () => eventTrigger.click(getElement(element)),
+  targetText: () => getElement(element).querySelector('span').innerHTML,
```
[the remark this answers is not in the document]
Comment (amiryonatan) [reply packages/wix-ui-core/src/components/Dropdown/Dropdown.driver.ts]: 👍

File: packages/wix-ui-core/src/components/Dropdown/Dropdown.spec.tsx [hunk 23]
```diff
@@ -0,0 +1,145 @@
...
+  });
+
+  describe('onSelect', () => {
+    it('should be called when selection changed', () => {
+      const onSelect = jest.fn();
+      const driver = createDriver(createDropdown({options, onSelect}));
+
+      driver.click();
+      driver.clickOptionAt(0);
+      expect(onSelect).toHaveBeenCalled();
```
[the remark this answers is not in the document]
Comment (amiryonatan) [reply packages/wix-ui-core/src/components/Dropdown/Dropdown.spec.tsx]: 👍

File: packages/wix-ui-core/src/components/Dropdown/Dropdown.spec.tsx [hunk 24]
```diff
@@ -0,0 +1,145 @@
...
+    });
+
+    it('should trigger onDeselect', () => {
+      const onDeselect = jest.fn();
+      const driver = createDriver(createDropdown({options, mode: MULTI_SELECT, onDeselect}));
+
+      driver.click();
+      driver.clickOptionAt(0);
+      driver.clickOptionAt(0);
+      expect(onDeselect).toHaveBeenCalled();
```
[the remark this answers is not in the document]
Comment (amiryonatan) [reply packages/wix-ui-core/src/components/Dropdown/Dropdown.spec.tsx]: 👍

File: packages/wix-ui-core/src/components/IconWithOptions/IconWithOptions.tsx (hunk 9 above)
Comment (amiryonatan) [reply packages/wix-ui-core/src/components/IconWithOptions/IconWithOptions.tsx]: 👍

File: packages/wix-ui-core/src/components/IconWithOptions/IconWithOptions.tsx (hunk 7 above)
Comment (amiryonatan) [reply packages/wix-ui-core/src/components/IconWithOptions/IconWithOptions.tsx]: 👍

**Pull request merged**